### PR TITLE
PLA-6199 design workflows

### PR DESCRIPTION
**Pull request by talltea**
# Citrine Python PR

## Description 
wip of adding Design Workflows/Executions 

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [ ] I have added tests for 100% coverage
- [ ] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [ ] I have bumped the version in setup.py

### CODE REVIEW

File: src/citrine/informatics/design_candidate.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    category_probability = properties.Mapping(properties.String, properties.Float, 'cp')
+
+    def __init__(self):
+        pass  # pragma: no cover
+
+
+class Mixture(Serializable["Mixture"], DesignVariable):
+    """docstring for MeanAndStd"""
+
+    category_probability = properties.Mapping(properties.String, properties.Float, 'cp')
```
Comment (maxhutch) [inline src/citrine/informatics/design_candidate.py:59]: This looks copy/pasted and needs to be updated.

File: src/citrine/informatics/design_candidate.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    std = properties.Float('s')
+
+    def __init__(self):
+        pass  # pragma: no cover
+
+
+class CategoriesAndProbabilities(Serializable["CategoriesAndProbabilities"], DesignVariable):
+    """docstring for MeanAndStd"""
+
+    category_probability = properties.Mapping(properties.String, properties.Float, 'cp')
```
Comment (maxhutch) [inline src/citrine/informatics/design_candidate.py:50]: Is this a map or a pair of coordinated lists in the API?

File: src/citrine/informatics/workflows/design_workflow.py [hunk 3]
```diff
@@ -75,3 +76,77 @@ def executions(self) -> WorkflowExecutionCollection:
...
+    @property
+    def design_executions(self) -> DesignExecutionCollection:
+        """Return a resource representing all visible executions of this workflow."""
+        if getattr(self, 'project_id', None) is None:
+            raise AttributeError('Cannot initialize execution without project reference!')
+        return DesignExecutionCollection(
+            project_id=self.project_id, session=self.session, workflow_id=self.uid)
+
+
+class NewDesignWorkflow(Resource['NewDesignWorkflow'], Workflow):
```
Comment (maxhutch) [inline src/citrine/informatics/workflows/design_workflow.py:89]: What, if anything, is different between this class and the original DesignWorkflow?

File: src/citrine/resources/design_execution.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    workflow_id: str
+        Unique identifier of the workflow that was executed
+
+    """
+
+    _response_key = None
+
+    uid: UUID = properties.UUID('id', serializable=False)
+    workflow_id = properties.UUID('workflow_id', serializable=False)
+    status = properties.Optional(properties.String(), 'status', serializable=False)
```
Comment (maxhutch) [inline src/citrine/resources/design_execution.py:33]: I think we're missing `status_description` (the minor status).

File: src/citrine/resources/design_execution.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        path = self._get_path()
+        data = self.session.post_resource(path, {'score': execution_input.dump()})
+        self._check_experimental(data)
+        return self.build(data)
+
+    def _build_candidates(self, subset_collection: Iterable[dict]) -> Iterable[DesignCandidate]:
+        for candidate in subset_collection:
+            yield DesignCandidate.build(candidate)
+
+    def candidates(self,
```
Comment (maxhutch) [inline src/citrine/resources/design_execution.py:110]: Does this duplicate the functionality of DesignWorkflow.results?

File: src/citrine/resources/design_execution.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            Resources in this collection.
+
+        """
+        return self._paginator.paginate(page_fetcher=self._fetch_page,
+                                        collection_builder=self._build_collection_elements,
+                                        page=page,
+                                        per_page=per_page)
+
+    def delete(self, uid: Union[UUID, str]) -> Response:
+        """Design Workflow Executions cannot be deleted; they can be archived instead."""
```
Comment (maxhutch) [inline src/citrine/resources/design_execution.py:189]:
```suggestion
        """Design Workflow Executions cannot be deleted or archived."""
```

**Review by maxhutch — COMMENTED** [review]
Round of comments

File: src/citrine/informatics/design_candidate.py (hunk 2 above)
Comment (talltea) [reply src/citrine/informatics/design_candidate.py:50]: It's a map in the API: `cp: Map[String, Double]` in the DesignResponseObject: https://github.com/CitrineInformatics/platform-backend/blob/c74f9b98d6f8fb6e88252a35eae23c46ba7cc506/code/orion/ml-config-core/src/main/scala/io/citrine/orion/core/domain/DesignResultResponse.scala#L7-L13

File: src/citrine/informatics/design_candidate.py (hunk 1 above)
Comment (talltea) [reply src/citrine/informatics/design_candidate.py:59]: woops, had this fixed locally but didn't commit

**Comment by maxhutch** [comment]
https://jenkins.corp.citrine.io/job/nextgen-ITs/11545/ is against this branch

**Review by maxhutch — APPROVED** [review]
LGTM!

File: src/citrine/resources/design_execution.py (hunk 5 above)
Comment (talltea) [reply src/citrine/resources/design_execution.py]: removed

File: src/citrine/resources/design_execution.py (hunk 4 above)
Comment (talltea) [reply src/citrine/resources/design_execution.py:37]: fixing

**Review by maxhutch — APPROVED** [review]

**Pull request merged**